node/tools/src/k8s.rs

```diff
@@ -1,43 +1,82 @@
 use crate::{config, NodeAddr};
-use anyhow::{anyhow, Context};
-use k8s_openapi::api::{
-    apps::v1::Deployment,
-    core::v1::{Namespace, Pod},
+use anyhow::{anyhow, ensure, Context};
+use k8s_openapi::{
+    api::{
+        apps::v1::{Deployment, DeploymentSpec},
+        core::v1::{Container, Namespace, Pod, PodSpec, PodTemplateSpec},
+    },
+    apimachinery::pkg::apis::meta::v1::LabelSelector,
 };
 use kube::{
     api::{ListParams, PostParams},
-    core::ObjectList,
+    core::{ObjectList, ObjectMeta},
     Api, Client, ResourceExt,
 };
 use serde_json::json;
-use std::collections::HashMap;
+use std::{collections::HashMap, net::SocketAddr};
 use tokio_retry::strategy::FixedInterval;
 use tokio_retry::Retry;
 use tracing::log::info;
 use zksync_protobuf::serde::Serde;
 
+/// Docker image name for consensus nodes.
+const DOCKER_IMAGE_NAME: &str = "consensus-node";
+
+/// K8s namespace for consensus nodes.
+pub const DEFAULT_NAMESPACE: &str = "consensus";
+
 /// Get a kube client
 pub async fn get_client() -> anyhow::Result<Client> {
     Ok(Client::try_default().await?)
 }
 
-/// Get a kube client
-pub async fn get_consensus_node_ips(client: &Client) -> anyhow::Result<Vec<String>> {
-    let pods: Api<Pod> = Api::namespaced(client.clone(), "consensus");
+/// Get the IP addresses and the exposed port of the RPC server of the consensus nodes in the kubernetes cluster.
+pub async fn get_consensus_nodes_address(client: &Client) -> anyhow::Result<Vec<SocketAddr>> {
+    let pods: Api<Pod> = Api::namespaced(client.to_owned(), DEFAULT_NAMESPACE);
     let lp = ListParams::default();
-    let pod = pods.list(&lp).await?;
-    let a: Vec<String> = pod
+    let pods = pods.list(&lp).await?;
+    ensure!(
+        !pods.items.is_empty(),
+        "No consensus pods found in the k8s cluster"
+    );
+    let pod_addresses: Vec<SocketAddr> = pods
         .into_iter()
-        .filter(|pod| {
-            pod.clone()
-                .metadata
-                .name
-                .unwrap()
-                .starts_with("consensus-node")
+        .filter_map(|pod| {
+            let pod_spec = pod.spec.context("Failed to get pod spec").ok()?;
+            let pod_running_container = pod_spec
+                .containers
+                .first()
+                .context("Failed to get pod container")
```

Comment:
Okay, you're correct that returning a `Result` from this function is no longer useful when using `filter_map`. I made the change using `filter` followed by `map`, but the errors that arise in the `filter` closure are still being ignored because it requires a boolean and using `?` to return a `Result` is not possible. Let me know if this is acceptable. In the `map` closure, the errors should now be handled correctly.

Reply:
if you need to return an error from filter, then just use a regular loop over the elements instead.

Reply:
I ended up including everything in the for loop instead of having the for loop only for filtering and then mapping. I chose this solution mainly because:

- I can reuse some pod variables instead of retrieving them twice, once for filtering and then again for mapping and address parsing.
- This way, I only iterate over all the pods once. With the separate map approach, I had to create an empty vector, push all the filtered pods into it, and then convert it to an iterator for the address parsing part.

Let me know what you think.